From 17f848a30c4b92fc0bd9c8bf8f116a143da3fe94 Mon Sep 17 00:00:00 2001
From: "karthikgk@maxlinear.com" <karthikgk@lvdocker7.lvsw.maxlinear.com>
Date: Tue, 15 Jul 2025 04:17:45 -0700
Subject: [PATCH] WLANRTSYS-91787 : MLD Triband - Driver FW interface
 implementation.

Task : Initial changes needed between hostapd and driver interface ,driver and firmware to support 3 Link MLD creation.

signed-of-by : karthikgk <karthikgk@maxlinear.com>
---
 src/ap/mxl_mld.c              | 9 +++++----
 src/common/vendor_cmds_copy.h | 4 +++-
 2 files changed, 8 insertions(+), 5 deletions(-)

diff --git a/src/ap/mxl_mld.c b/src/ap/mxl_mld.c
index 71e2cd72d..0c549c736 100644
--- a/src/ap/mxl_mld.c
+++ b/src/ap/mxl_mld.c
@@ -303,19 +303,20 @@ static int mxl_setup_link(struct p_ap_mld *mld)
 		mld_info->mld_id = link1->conf->mld_id;
 		memcpy_s(mld_info->link1_bssid, ETH_ALEN, link1->conf->bssid, ETH_ALEN);
 		memcpy_s(mld_info->ap_mld_mac, ETH_ALEN, link1->conf->mxl_bss_conf.ap_mld_mac, ETH_ALEN);
+		mld_info->num_of_sibling_vaps = 1; /* Need to be Updated for 3 LINK MLD */
 
 		if (is_6ghz_op_class(link2->iconf->op_class))
-			mld_info->link2_vapid = (link2->conf->bss_idx -1) + VAPID_OFFSET_6G;
+			mld_info->sibling_vap_ids[0] = (link2->conf->bss_idx -1) + VAPID_OFFSET_6G;
 		else if (link2->iface->freq >= 5180 && link2->iface->freq <= 5900)
-			mld_info->link2_vapid = (link2->conf->bss_idx -1) + VAPID_OFFSET_5G;
+			mld_info->sibling_vap_ids[0] = (link2->conf->bss_idx -1) + VAPID_OFFSET_5G;
 		else
-			mld_info->link2_vapid = (link2->conf->bss_idx -1) + VAPID_OFFSET_2G;
+			mld_info->sibling_vap_ids[0] = (link2->conf->bss_idx -1) + VAPID_OFFSET_2G;
 
 		mld_link_info->mld_id = link1->conf->mld_id;
 		memcpy_s(mld_link_info->ap_mld_mac, ETH_ALEN, link1->conf->mxl_bss_conf.ap_mld_mac, ETH_ALEN);
 		wpa_printf(MSG_DEBUG, "mxl_setup_link for mld_id %d mld_mac="MACSTR"link1_bssid ="MACSTR"link2_vapid= %d",
 				mld_info->mld_id, MAC2STR(mld_info->ap_mld_mac),
-				MAC2STR(mld_info->link1_bssid), mld_info->link2_vapid);
+				MAC2STR(mld_info->link1_bssid), mld_info->sibling_vap_ids[0]);
 	}
 
 	/* Initialize wpa encryption parameters for the MLD VAPs */
diff --git a/src/common/vendor_cmds_copy.h b/src/common/vendor_cmds_copy.h
index 7fbc65643..2f5c58062 100644
--- a/src/common/vendor_cmds_copy.h
+++ b/src/common/vendor_cmds_copy.h
@@ -655,12 +655,14 @@ struct _mxl_vendor_bcast_twt_config_t {
 #define LINK_ID_6G                2
 #define LINK_ID_5G                1
 #define LINK_ID_2G                0
+#define NUM_OF_SIBLING_LINKS      2
 
 struct _mxl_vendor_mld_info {
   int mld_id;
   u8 ap_mld_mac[ETH_ALEN];
   u8 link1_bssid[ETH_ALEN];
-  u8 link2_vapid;
+  u8 num_of_sibling_vaps;
+  u8 sibling_vap_ids[NUM_OF_SIBLING_LINKS];
 } __attribute__ ((packed));
 
 struct _mxl_vendor_mld_link_info {
-- 
2.43.0

